beaconclient/multi_beacon_client.go

```diff
@@ -0,0 +1,184 @@
+// Package beaconclient provides a beacon-node client
+package beaconclient
+
+import (
+	"context"
+	"errors"
+	"os"
+	"sync"
+
+	"github.com/flashbots/go-boost-utils/types"
+	"github.com/sirupsen/logrus"
+	uberatomic "go.uber.org/atomic"
+)
+
+var (
+	ErrBeaconNodeSyncing      = errors.New("beacon node is syncing")
+	ErrBeaconNodesUnavailable = errors.New("all beacon nodes responded with error")
+)
+
+// IMultiBeaconClient is the interface for the MultiBeaconClient, which can manage several beacon client instances under the hood
+type IMultiBeaconClient interface {
+	BestSyncStatus() (*SyncStatusPayloadData, error)
+	SubscribeToHeadEvents(slotC chan HeadEventData)
+	FetchValidators(headSlot uint64) (map[types.PubkeyHex]ValidatorResponseEntry, error)
+	GetProposerDuties(epoch uint64) (*ProposerDutiesResponse, error)
+}
+
+// IBeaconInstance is the interface for a single beacon client instance
+type IBeaconInstance interface {
+	SyncStatus() (*SyncStatusPayloadData, error)
+	CurrentSlot() (uint64, error)
+	SubscribeToHeadEvents(slotC chan HeadEventData)
+	FetchValidators(headSlot uint64) (map[types.PubkeyHex]ValidatorResponseEntry, error)
+	GetProposerDuties(epoch uint64) (*ProposerDutiesResponse, error)
+	GetURI() string
+}
+
+type MultiBeaconClient struct {
+	log             *logrus.Entry
+	bestBeaconIndex uberatomic.Int64
+	beaconInstances []IBeaconInstance
+
+	// feature flags
+	ffAllowSyncingBeaconNode bool
+}
+
+func NewMultiBeaconClient(log *logrus.Entry, beaconInstances []IBeaconInstance) *MultiBeaconClient {
+	client := &MultiBeaconClient{
+		log:             log.WithField("module", "beaconClient"),
+		beaconInstances: beaconInstances,
+		bestBeaconIndex: *uberatomic.NewInt64(0),
+	}
+
+	// feature flags
+	if os.Getenv("ALLOW_SYNCING_BEACON_NODE") != "" {
+		client.log.Warn("env: ALLOW_SYNCING_BEACON_NODE: allow syncing beacon node")
+		client.ffAllowSyncingBeaconNode = true
+	}
+
+	return client
+
+}
+
+func (c *MultiBeaconClient) BestSyncStatus() (*SyncStatusPayloadData, error) {
+	var bestSyncStatus *SyncStatusPayloadData
+	var numSyncedNodes uint32
+	requestCtx, requestCtxCancel := context.WithCancel(context.Background())
```

Comment:
that requestCtx is not used as context, just to not count additional responses 🤔 

does this matter?